api/dataset.go

```diff
@@ -2,38 +2,218 @@ package api
 
 import (
 	"context"
+	"errors"
 	"net/http"
+	"strings"
+
+	"github.com/ONSdigital/dp-api-clients-go/v2/cantabular"
+	"github.com/ONSdigital/dp-api-clients-go/v2/dataset"
+	"github.com/go-chi/chi/v5"
 )
 
-func (api *API) getDatasetJSON(w http.ResponseWriter, r *http.Request) {
-	var response *getDatasetJsonObservationsResponse
-	var err error
+type optionsMap map[string]map[string]dataset.Option
+
+type datasetParams struct {
+	id                 string
+	edition            string
+	version            string
+	basedOn            string
+	unsortedDimensions []string
+	options            optionsMap // dimension -> option -> option item
+}
 
+func (api *API) getDatasetJSON(w http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
 
 	if r.URL.Query().Get("geography") == "" {
-		response, err = api.getDefaultDatasetJSON(ctx, r)
+		api.getDefaultDatasetJSON(ctx, w, r)
 	} else {
-		response, err = api.getGeographyDatsetJSON(ctx, r)
+		api.getGeographyDatsetJSON(ctx, r)
+	}
+}
+
+func (api *API) getDefaultDatasetJSON(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	params, err := api.getDatasetParams(ctx, r)
+
+	if err != nil {
+		api.respond.Error(ctx, w, http.StatusNotFound, err)
+		return
+	}
+
+	geoDimensions, err := api.getGeographyTypes(ctx, params.basedOn)
+	if err != nil {
+		api.respond.Error(ctx, w, http.StatusInternalServerError, err)
```

Comment:
Please don't just return status.InternalServerError, we have set up a system that can reflexively determine statusCodes from information embedded in the errors. As a default you should return statusCode(err), which will attempt to determine the statusCode and default to status 500. Here the error returned by cantabular will return the actual status code. Other status codes are achieved by returning errors that conform to interfaces found in package `errors`